### PR TITLE
Article/esp32 memory map 101

### PR DESCRIPTION
**Pull request by marekmatej**
Repost & update the article from Medium.

### CODE REVIEW

**Comment by marekmatej** [comment]
@f-hollow, can you PTAL

**Comment by f-hollow** [comment]
@marekmatej 

Thank you for your contribution!

Please rebase this article on main and make sure that all required checks pass.

**Comment by pedrominatel** [comment]
Thanks, @marekmatej 

Currently, we are working on the Medium content migration. Just a reminder here to not duplicate this article during the migration. cc @georgik

File: content/blog/esp32-memory-map-101/index.md [hunk 1]
```diff
@@ -0,0 +1,351 @@
...
+tags: ["ESP32", "ESP32-S2", "ESP32-S3", "ESP32-C3", "ESP32-C6"]
+showAuthor: false
+authors:
+  - "marek-matej"
+---
+
+
+The performance and memory capabilities of the microcontrollers are increasing, and so is their complexity, especially when the users are responsible for mapping external memory chips such as Flash or SPIRAM by configuring a memory management unit.
+
+When I started to use Espressif ESP32’s series SoC’s with the [Zephyr RTOS](https://docs.zephyrproject.org/latest/index.html), quickly I realized that these SoC’s are a very different world, compared to my previous MCU experiences, which were mostly ARM-related. ARM Cortex-M’s - in most cases - uses Von-Neumann memory architecture, while the ESP32’s series of microcontrollers uses the Harvard memory architecture, which adds complexity.
```
Comment (f-hollow) [inline content/blog/esp32-memory-map-101/index.md:13]: Regarding ESP32's series of SoC's:

First of all, the abbreviation SoCs is more common in Espressif documentation. To my surprise, Wikipedia, also [suggests](https://en.wikipedia.org/wiki/System_on_a_chip) this plural form.

There is a number of ways of how a subset of chips is referred to in Espressif documentation and articles: (1) ESP32 series of SoCs, (2) ESP32 family of SoCs, etc. Sometimes these terms are interchangeable. Sometimes, they are used side by side with modified meanings. At least in the hardware documentation, we started removing family wherever possible. Suggest that you also use ESP32 series of SoCs.

In summary, please consider using _ESP32 series of SoCs_.

File: content/blog/esp32-memory-map-101/index.md [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+The primary source of knowledge about any SoC was always the datasheet or Technical Reference Manual of the part. Indeed it still is, but I found the ESP32's TRM — although very useful — quite challenging to read. Especially because there is a significant portion of important information buried in the long textual descriptions.
```
Comment (f-hollow) [inline content/blog/esp32-memory-map-101/index.md:15]: Very true! it is also quite challenging to structure or edit this documentation :)

File: content/blog/esp32-memory-map-101/index.md [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+
+This sparked the motivation to compile a visually rich and informative depiction of crucial information from the current versions of a TRM. This is intended to be used together with the SoC’s datasheets and TRM to help with a better understanding of the ESP32's family SoCs internals, which is crucial during the system and driver development - for example when porting to other platforms such as Zephyr RTOS.
```
Comment (f-hollow) [inline content/blog/esp32-memory-map-101/index.md:17]:
```suggestion
This sparked the motivation to compile a visually rich and informative depiction of crucial information from the current versions of a TRM. This is intended to be used together with the SoC’s datasheets and TRM to help better understand the internals of the ESP32 series of SoCs, which is crucial during the system and driver development - for example when porting to other platforms such as Zephyr RTOS.
```

File: content/blog/esp32-memory-map-101/index.md [hunk 4]
```diff
@@ -0,0 +1,351 @@
...
+
+## ESP32
+
+The ESP32 is the Xtensa LX6 dual-core microcontroller, and it is the oldest SoC from the ESP32’s family of MCUs.
+
+The SoC has access to external Flash memory using the SPI interface. It can also access the external SPIRAM, but only on the data bus.
```
Comment (f-hollow) [inline content/blog/esp32-memory-map-101/index.md:23]: There is nothing special to our flash memory to make it _Flash_. Suggest using _flash memory_ without capitalization.

File: content/blog/esp32-memory-map-101/index.md [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+
+Zephyr-RTOS support for the ESP32 devkit board can be found [here](https://docs.zephyrproject.org/latest/boards/espressif/esp32_devkitc_wrover/doc/index.html).
+
+
+**ROM**
+
+ROM memories hold the primary bootloader code and other library functions available to the end users. The content of the ROM memories is baked during the manufacturing process and cannot be changed. Please refer to the ECO versions of the ESP32 microcontroller for more detailed information. TLDR is to use the last ECOv3 and later ESP32 chip versions.
```
Comment (f-hollow) [inline content/blog/esp32-memory-map-101/index.md:30]: ECO is a term referring to an internal document describing the changes. Lately, there was an initiative to scrape it off release notes in some projects.

The suggested replacement is _chip revision_. The chip revisions are documented in the public document called [Errata](https://www.espressif.com/sites/default/files/documentation/esp32_errata_en.pdf#page=6).

File: content/blog/esp32-memory-map-101/index.md [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+RTC-Slow memory is r/w memory that can be accessed by either CPU via the instruction and data bus.
+
+The RTC memories can hold the data during the deep sleep period.
+
+**Instruction & data cache**
+
+The instruction and data cache address space is a virtual address space where the external memories such as Flash or SPIRAM modules can be accessed by the instruction or the data bus.
+
+
+### Internal memories and cache address space
```
Comment (f-hollow) [inline content/blog/esp32-memory-map-101/index.md:53]: My first reaction to this image is that it reminds me of Figure 1-2. System Address Mapping in the [TRM](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf#page=25).

Since you are "decoding" the information from the TRM in this article, do you think it is worth mentioning this image here and contrasting or comparing them?

Feel free to disregard this idea if you don't see much value.

File: content/blog/esp32-memory-map-101/index.md [hunk 7]
```diff
@@ -0,0 +1,351 @@
...
+    default=true
+    src="img/esp32-periph.webp"
+    alt=""
+    caption="The ESP32 peripherals map."
+    >}}
+
+
+### e-Fuse blocks
+
+The eFuse memory is the OTP memory where hardware and encryption settings are stored. Fundamentally, an eFuse is a single bit of non-volatile memory with the restriction that once an eFuse bit is programmed to 1, it can never be reverted to 0.
```
Comment (f-hollow) [inline content/blog/esp32-memory-map-101/index.md:77]: The abbreviation OTP is new to me. Not sure if I saw it in Epsressif docs. There is likelihood that it can be new to readers as well.

Suggest that you expand OTP and, if it is not self-explanatory, maybe explain what it is.

File: content/blog/esp32-memory-map-101/index.md [hunk 8]
```diff
@@ -0,0 +1,351 @@
...
+
+
+For further information and all the details about the ESP32-S3 SoC please refer to the latest [ESP32-S3 Technical Reference Manual](https://www.espressif.com/sites/default/files/documentation/esp32-s3_technical_reference_manual_en.pdf).
+
+---
+
+
+## ESP32-C3
+
+The ESP32-C3 is a 32-bit RISC-V microcontroller with the “IMC” extensions. The address space of ESP32-C3 is significantly simpler compared to its Xtensa counterparts. It supports JTAG debugging via the on-chip USB interface.
```
Comment (f-hollow) [inline content/blog/esp32-memory-map-101/index.md:219]: The abbreviation IMC seems unfamiliar to me as well. There is likelihood that it can be new to readers.

Suggest that you expand it and, if it is not self-explanatory, maybe explain what it is.

File: content/blog/esp32-memory-map-101/index.md [hunk 9]
```diff
@@ -0,0 +1,351 @@
...
+    alt=""
+    caption="The ESP32 memory map."
+    >}}
+
+
+### Peripherals
+
+The ESP32 has 41 peripherals. Nearly all peripheral modules can be accessed by either CPU at the same address with a single exception of the PID controller which exists for each CPU and can be accessed only by its own CPU.
+
+{{< figure
```
Comment (f-hollow) [inline content/blog/esp32-memory-map-101/index.md:67]: If I understand it correctly, the listed regions of memory contain registers for the respective peripherals.

Since you are deciphering TRM content here, and with the fact that registers are a big part of a TRM, suggest that you mention registers here.

**Review by f-hollow — APPROVED** [review]
@marekmatej

Thank you for contributing the article!

A few nitpicks from me. If you have time now, please implement my comments. If not, feel free to resolve them, but take these points into account for the future.

File: content/blog/esp32-memory-map-101/index.md (hunk 6 above)
Comment (marekmatej) [reply content/blog/esp32-memory-map-101/index.md:53]: You are right, the fig1.2 is the equivalent picture of the memory map, but I would like to leave it to the reader to figure this out. Bringing out a contrast with the TRM is not what I want to focus in this article, and TBH I think it could lead to confusion.

File: content/blog/esp32-memory-map-101/index.md (hunk 8 above)
Comment (marekmatej) [reply content/blog/esp32-memory-map-101/index.md:219]: Since I wasn't able to find a source of documentation on the RISC-V extension that I could link as an explanation, I removed the reference altogether. It is not important in the context of the article.

File: content/blog/esp32-memory-map-101/index.md (hunk 7 above)
Comment (marekmatej) [reply content/blog/esp32-memory-map-101/index.md:77]: Fixed

File: content/blog/esp32-memory-map-101/index.md (hunk 5 above)
Comment (marekmatej) [reply content/blog/esp32-memory-map-101/index.md:30]: Fixed

File: content/blog/esp32-memory-map-101/index.md (hunk 4 above)
Comment (marekmatej) [reply content/blog/esp32-memory-map-101/index.md:23]: fixed

File: content/blog/esp32-memory-map-101/index.md (hunk 3 above)
Comment (marekmatej) [reply content/blog/esp32-memory-map-101/index.md:17]: I think both "family" and "series" are used broadly, however, I'm using the term "series" inside the article but leave the title of the article as is, with the term "family". (This is how it was published on Medium)

File: content/blog/esp32-memory-map-101/index.md (hunk 2 above)
Comment (marekmatej) [reply content/blog/esp32-memory-map-101/index.md:15]: ...and the main motivation for this article as well :)

**Comment by f-hollow** [comment]
@marekmatej Thank you, LGTM!
